### PR TITLE
Hotfix: Filtering Space from DataCite Resource Types at Export and fix CI check for PHPUnit Tests

### LINKED ISSUE #1182: Failed generation of valid DataCite XML when users select resource types "Computational Notebook", "Data Paper" or "Interactive Resource"
### Description

ELMO currently fails to generate valid DataCite XML when users select the resource types "Computational Notebook", "Data Paper" or "Interactive Resource".  The "resourceTypeGeneral" attribute is written into XML containing spaces. 

Image

### Steps to Reproduce

1. Open ELMO - GFZ Metadata Editor  /elmo
2. Fill required fields with test information 
3. On the dropdown Resource Information fill "Resource type" with  "Computational Notebook", "Data Paper" or "Interactive Resource"
4. Click on "Save as XML" 
5. Open saved XML file 
6. Check field `resourceTypeGeneral`

### Error(s)

```shell

```

### Expected Behavior

ELMO has to generate valid DataCite XML when users select the resource types "Computational Notebook", "Data Paper" or "Interactive Resource".  The "resourceTypeGeneral" attribute written into XML should not contain spaces and should be written in a way:
DataPaper
ComputationalNotebook
InteractiveResource

### Screenshots / Logs

_No response_

### Browser

Mozilla Firefox 140.13.0 ESR

### ELMO Version

Version 2.1.1. - 2026/07/17

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by McNamara84**
This pull request introduces several improvements and bug fixes across the codebase, focusing on more robust handling of author and spatial/temporal coverage form data, improved resource type matching in the UI, and enhanced test validation in CI. These changes address edge cases with empty or partial form submissions, ensure more accurate data processing, and improve developer feedback during test runs.

### Form Data Handling and Validation
* Refactored `filterValidPersonAuthors` in `save_authors.php` to handle partial and missing author fields more robustly, ensuring that any row with at least one non-empty name part or ORCID is preserved, and that all author-related arrays are handled safely even if missing or of different lengths. Validation now consistently uses this logic. [[1]](diffhunk://#diff-06893c4e5a015052d3aead31fcdcc17a7d6bc40c4b51b399554ad842b12d70b5L31-R50) [[2]](diffhunk://#diff-06893c4e5a015052d3aead31fcdcc17a7d6bc40c4b51b399554ad842b12d70b5L76-R73) [[3]](diffhunk://#diff-06893c4e5a015052d3aead31fcdcc17a7d6bc40c4b51b399554ad842b12d70b5L284-R269)
* Rewrote `save_spatialtemporalcoverage.php` to better handle sparse or partial spatial/temporal rows, only saving rows with actual data (ignoring default-only timezone), and enforcing required field dependencies on submit while allowing drafts to save partial data.
* 
### UI and Data Mapping Improvements
* Added `normalizeResourceTypeGeneral` and `findResourceTypeOption` utility functions to both `doiPrefill.js` and `mappingXmlToInputFields.js`, improving resource type selection from DataCite/ERNIE by supporting both exact and whitespace-insensitive label matches. Updated resource type prefill and XML mapping to use these helpers. [[1]](diffhunk://#diff-0712b629677cce247549fcb647ca8c51576547bf86a1aff5c5c064542a23256fR220-R244) [[2]](diffhunk://#diff-0712b629677cce247549fcb647ca8c51576547bf86a1aff5c5c064542a23256fL248-R276) [[3]](diffhunk://#diff-67c1fa1d0f6d5ed82d952e3b2cdacac9fa7f7565e11567a1224cae30c3ad4b85R8-R40) [[4]](diffhunk://#diff-67c1fa1d0f6d5ed82d952e3b2cdacac9fa7f7565e11567a1224cae30c3ad4b85L38-R75) [[5]](diffhunk://#diff-67c1fa1d0f6d5ed82d952e3b2cdacac9fa7f7565e11567a1224cae30c3ad4b85R1791-R1792)
* Updated spatial/temporal coverage validation in `checkMandatoryFields.js` to require all relevant fields when any temporal value is filled, and to properly distinguish between spatial-only, temporal-only, and description-only entries. [[1]](diffhunk://#diff-fdef90dcaa21f295b52068f84f0d6bc01b0fe5289e7cbda5e70884bab282be77L217-R233) [[2]](diffhunk://#diff-fdef90dcaa21f295b52068f84f0d6bc01b0fe5289e7cbda5e70884bab282be77L249-R257) [[3]](diffhunk://#diff-fdef90dcaa21f295b52068f84f0d6bc01b0fe5289e7cbda5e70884bab282be77R266-R272)

### Testing and CI Enhancements
* Improved the PHPUnit workflow in `.github/workflows/phpunit.yml` to require a valid JUnit report and nonzero test count, and to check for a proper test summary, ensuring test runs that terminate early or without executing tests are caught as failures.

### Infrastructure and Error Handling
* Updated `save_data.php` to better preserve output buffer state for callers (such as PHPUnit), and to avoid premature exits in non-POST or test environments, improving testability and error handling. [[1]](diffhunk://#diff-6b740abc2601dce399cf05575be39f85546d425e44ecdb17e9fff489247a2c9aL18-R25) [[2]](diffhunk://#diff-6b740abc2601dce399cf05575be39f85546d425e44ecdb17e9fff489247a2c9aL44-R47) [[3]](diffhunk://#diff-6b740abc2601dce399cf05575be39f85546d425e44ecdb17e9fff489247a2c9aL85-R83)

## Notes for Reviewer
None.

### Branching Strategy:
If this is a hotfix: The branch was created from `main` and will be merged into `main`.
If this is a feature or documentation change: The branch was created from `dev` and will be merged into `dev`.
If this is a feature or documentation branch, I have pulled the latest changes from `main` into my branch.

### Code Quality
- [x] My code follows the style guide.
- [x] I have self-reviewed my code.
- [x] I added comments for hard-to-understand code.
- [x] My changes do not create new warnings in the browser console.

### Documentation
- [x] If applicable, PHP code is documented using PHPDoc.
- [x] If applicable, JavaScript code is documented using JSDoc.
- [x] If needed, the ELMO Guide ('./doc/help.html') has been updated.
- [x] If needed, the README has been updated.
- [x] If needed, the API documentation ('./api/v2/docs') has been updated.
- [x] If a new feature was added or a bug fixed, the changelog ('./doc/changelog.html') has been updated.

### Testing
- [x] If needed, Playwright tests have been updated or added.
- [x] If needed, unit tests have been updated or added.
- [x] If needed, jest tests have been updated or added.

## Known Issues
None.

### CODE REVIEW

**Comment by Daetha** [comment]
I am on it

**Comment by Daetha** [comment]
Sorry, didn't mean to close the PR

**Comment by McNamara84** [comment]
I have time to fix problems trough my changes til 17:00pm...

File: js/doiPrefill.js [hunk 1]
```diff
@@ -217,6 +217,31 @@ function normalizeRelationType(relationType) {
...
+ * Prefers an exact option label and falls back to whitespace-independent matching.
+ * @param {HTMLOptionElement[]} options
+ * @param {string} resourceTypeGeneral
+ * @returns {HTMLOptionElement|undefined}
+ */
+function findResourceTypeOption(options, resourceTypeGeneral) {
+  const exactMatch = options.find(option => option.text.trim() === resourceTypeGeneral);
+  if (exactMatch) return exactMatch;
+
+  const normalizedResourceType = normalizeResourceTypeGeneral(resourceTypeGeneral);
```
Comment (Daetha) [inline js/doiPrefill.js]: ok, this is the solution to work with or without spaces on front-end

File: js/mappingXmlToInputFields.js [hunk 2]
```diff
@@ -5,6 +5,39 @@
...
  * @param {Function} resolver - The namespace resolver function.
  */
+/**
+ * Creates the whitespace-independent key used by DataCite and ERNIE labels.
+ *
+ * @param {unknown} value
+ * @returns {string}
+ */
+function normalizeResourceTypeGeneral(value) {
+  return String(value ?? '').replace(/\s+/g, '');
```
Comment (Daetha) [inline js/mappingXmlToInputFields.js]: same function defined twice... ok, let it be in a hotfix

**Comment by Daetha** [comment]
Dear @McNamara84, you have successfully modified the import and export paths, in JS and XSLT. 

The problem is however in the save part: **i was unable to save file with "Computational notebook" selected.**

```
web-1  | [Wed Jul 29 14:17:37.074399 2026] [php:notice] [pid 31:tid 31] [client 172.20.0.1:58422] Error creating new resource: Cannot add or update a child row: a foreign key constraint fails (`elmocache`.`Resource`, CONSTRAINT `Resource_ibfk_2` FOREIGN KEY (`Resource_Type_resource_name_id`) REFERENCES `Resource_Type` (`resource_name_id`)), referer: http://localhost:8080/
web-1  | [Wed Jul 29 14:17:37.074419 2026] [php:notice] [pid 31:tid 31] [client 172.20.0.1:58422] Error in saveResourceInformationAndRights: Cannot add or update a child row: a foreign key constraint fails (`elmocache`.`Resource`, CONSTRAINT `Resource_ibfk_2` FOREIGN KEY (`Resource_Type_resource_name_id`) REFERENCES `Resource_Type` (`resource_name_id`)), referer: http://localhost:8080/
web-1  | [Wed Jul 29 14:17:37.074422 2026] [php:notice] [pid 31:tid 31] [client 172.20.0.1:58422] [\xf0\x9f\x92\xbfSAVE]: Save operation failed: saveResourceInformationAndRights returned false, referer: http://localhost:8080/
web-1  | [Wed Jul 29 14:17:37.074425 2026] [php:notice] [pid 31:tid 31] [client 172.20.0.1:58422] [\xf0\x9f\x92\xbfSAVE]: Exception in saveResourceInformationAndRights: Save operation failed: saveResourceInformationAndRights returned false, referer: http://localhost:8080/
web-1  | [Wed Jul 29 14:17:37.074467 2026] [php:notice] [pid 31:tid 31] [client 172.20.0.1:58422] [SAVE] Transaction rolled back 
```

but the resourceTypes selection is remotely controlled by ERNIE:

`2026-07-29 14:17:09 [Wed Jul 29 14:17:09.290873 2026] [php:notice] [pid 27:tid 27] [client 172.20.0.1:62112] Description Types: Serving 6 types from ERNIE (cache or fresh), referer: http://localhost:8080/`

## Concept:
I agree with your vision that:

Import should import with or without spaces ✔️
Export should always remove spaces 
Storage (suggested) should save with spaces. Because this is how user sees the data and ERNIE serves the data.

## Problem:
Currently the database has no spaces on init, but then ERNIE comes with spaces aand tries to sync

suggested change:

```
        "Resource_Type" => [
            ["resource_type_general" => "Audiovisual", "description" => "A series of visual representations imparting an impression of motion when shown in succession. May or may not include sound."],
            ["resource_type_general" => "Collection", "description" => "An aggregation of resources, which may encompass collections of one resourceType as well as those of mixed types. A collection is described as a group; its parts may also be separately described."],
            ["resource_type_general" => "Computational Notebook", "description" => "A virtual notebook environment used for literate programming."],
            ["resource_type_general" => "Data Paper", "description" => "A factual and objective publication with a focused intent to identify and describe specific data, sets of data, or data collections to facilitate discoverability."],
            ["resource_type_general" => "Dataset", "description" => "Data encoded in a defined structure."],
            ["resource_type_general" => "Event", "description" => "A non-persistent, time-based occurrence."],
            ["resource_type_general" => "Image", "description" => "A visual representation other than text."],
            ["resource_type_general" => "Interactive Resource", "description" => "A resource requiring interaction from the user to be understood, executed, or experienced."],
            ["resource_type_general" => "Model", "description" => "An abstract, conceptual, graphical, mathematical or visualization model that represents empirical objects, phenomena, or physical processes."],
            ["resource_type_general" => "Output Management Plan", "description" => "A formal document that outlines how research outputs are to be handled both during a research project and after the project is completed."],
            ["resource_type_general" => "Preprint", "description" => "A version of a scholarly or scientific paper that precedes formal peer review and publication in a peer-reviewed scholarly or scientific journal."],
            ["resource_type_general" => "Software", "description" => "A computer program other than a computational notebook, in either source code (text) or compiled form. Use this type for general software components supporting scholarly research. Use the \"Computational Notebook\" value for virtual notebooks."],
            ["resource_type_general" => "Sound", "description" => "A resource primarily intended to be heard."],
            ["resource_type_general" => "Standard", "description" => "Something established by authority, custom, or general consent as a model, example, or point of reference."],
            ["resource_type_general" => "Study Registration", "description" => "A detailed, time-stamped description of a research plan, often openly shared in a registry or published in a journal before the study is conducted to lend accountability and transparency in the hypothesis generating and testing process."],
            ["resource_type_general" => "Text", "description" => "A resource consisting primarily of words for reading that is not covered by any other textual resource type in this list."],
            ["resource_type_general" => "Workflow", "description" => "A structured series of steps which can be executed to produce a final outcome, allowing users a means to specify and enact their work in a more reproducible manner."],
            ["resource_type_general" => "Other", "description" => "If selected, supply a value for ResourceType."],
            ["resource_type_general" => "Award", "description" => "A grant, prize, or other financial or honorary acknowledgment that supports or recognizes research contributions."],
            ["resource_type_general" => "Project", "description" => "An organized research activity with defined goals, timeline, and resources, often funded by grants or institutional support."]
        ],
```

## Question:
After a full reload I of the database was able to save resources. Result looks like:
`<resourceType resourceTypeGeneral="ComputationalNotebook">Computational Notebook</resourceType>`
 -  is it intended to have the XML attribute without spaces and value with spaces?
 - the upload of such file was success

**Review by Daetha — CHANGES_REQUESTED** [review]
Good job, just a bit more is needed to make it work!

Sooo, if i request changes anyway, will you also remove the duplicated function definitions in `js/mappingXmlToInputFields.js` and `js/doiPrefill.js` ?

**Comment by McNamara84** [comment]
> * is it intended to have the XML attribute without spaces and value with spaces?

Yes, it is. The value of the Element can theoretisch be something but the Attribute `resourceTypeGeneral` can only be a word from [the controlled vocab of the schema](https://datacite-metadata-schema.readthedocs.io/en/4.7/properties/resourcetype/).

**Comment by McNamara84** [comment]
The other points will be fixed with my next commit. I will re-request a review when I am ready with it.

**Comment by McNamara84** [comment]
All of your points should be fixed now. Outsourcing the util was a bit heavy because it is now in an own file registered in the footer. But it works for now and you were right - it was not DRY...

**Comment by McNamara84** [comment]
Oh f... Checks are red. Will look into the log... Damn...

**Review by Daetha — APPROVED** [review]
Everything left is test fixing, so I approve the core changes. After the tests are fixed or the breaking changes reverted, merge can start.

Best regards,
Sasha
